velox/functions/lib/Re2Functions.cpp

```diff
@@ -116,6 +116,48 @@ bool re2Extract(
   }
 }
 
+void checkEscape(bool condition) {
+  if (!condition) {
+    VELOX_USER_FAIL(
+        "Escape character must be followed by '%%', '_' or the escape character itself");
+  }
+}
+
+StringView likePatternToRe2(StringView pattern, char escapeChar) {
+  std::string regex;
+  regex.append("^");
+  bool escaped = false;
+  for (const char* c = pattern.begin(); c < pattern.end(); ++c) {
+    checkEscape(!escaped || *c == '%' || *c == '_' || *c == escapeChar);
+    if (!escaped && (*c == escapeChar)) {
+      escaped = true;
+    } else {
+      switch (*c) {
+        case '%':
+          regex.append(escaped ? "%" : ".*");
+          escaped = false;
+          break;
+        case '_':
+          regex.append(escaped ? "_" : ".");
+          escaped = false;
+          break;
+        case '\\':
```

Comment:
Are you trying to catch all regex metacharacters that need escaping here?
If so, you need at least '[]'. I'd check the re2 parser to ensure you've covered all the relevant cases:

https://github.com/google/re2/blob/main/re2/parse.cc#L2262

I think you need at least these characters: `|{}()[]`

Reply:
Here is relevant code from Presto: presto-main/src/main/java/com/facebook/presto/type/LikeFunctions.java

```
    private static Regex likePattern(String patternString, char escapeChar, boolean shouldEscape)
    {
        StringBuilder regex = new StringBuilder(patternString.length() * 2);

        regex.append('^');
        boolean escaped = false;
        for (char currentChar : patternString.toCharArray()) {
            checkEscape(!escaped || currentChar == '%' || currentChar == '_' || currentChar == escapeChar);
            if (shouldEscape && !escaped && (currentChar == escapeChar)) {
                escaped = true;
            }
            else {
                switch (currentChar) {
                    case '%':
                        regex.append(escaped ? "%" : ".*");
                        escaped = false;
                        break;
                    case '_':
                        regex.append(escaped ? "_" : ".");
                        escaped = false;
                        break;
                    default:
                        // escape special regex characters
                        switch (currentChar) {
                            case '\\':
                            case '^':
                            case '$':
                            case '.':
                            case '*':
                                regex.append('\\');
                        }

                        regex.append(currentChar);
                        escaped = false;
                }
            }
        }
        checkEscape(!escaped);
        regex.append('$');

        byte[] bytes = regex.toString().getBytes(UTF_8);
        return new Regex(bytes, 0, bytes.length, Option.MULTILINE, NonStrictUTF8Encoding.INSTANCE, SYNTAX);
    }
```

Reply:
The characters that need escaping is dependent on the metacharacters in the regex engine being used. Is this Presto code translating to an RE2 regex?

Since RE2 is the underlying regex engine here, IMO the RE2 parser is the best source for a list of metacharacters.